모임 삭제 API 구현

hyeonnkang opened the pull request:
## 작업 사항
close #49
- 모임을 삭제 하는 API를 컨트롤러 Layer 까지 생성하고 swagger으로 삭제 테스트 까지 진행했습니다
- 모임 단건 조회 구현 할 때 Meet 의 toString 코드가 develop 브랜치에 잘 못 들어갔는데 이 브랜치 merge 할 때 toString 코드 없앨지 의견 주시면 감사하겠습니다!
- 
현재 mapping 메소드가 DELETE로 되어있는데 추가 테이블에 컬럼을 넣고 PATCH로 변경할 지는 추후 논의가 필요할 것 같아요!

## 참고 자료

- 참고자료 (링크 등)

## 체크리스트

- [x]  PR의 제목은 팀 내 규칙을 준수하여 알맞게 작성하였는가?
- [x]  Merge 하는 브랜치가 올바른가? (`main` branch에 실수로 PR 생성 금지)
- [x]  팀의 코딩 컨벤션을 준수하는가?
- [x]  PR과 관련없는 변경사항이 들어가지는 않았는가?
- [x]  내 코드에 대한 자기 검토가 되었는가?
- [x]  Reviewers, Assignees, Lables, Project, Milestone은 적절하게 선택하였는가?
- [x]  관련한 issue를 닫아야 하는지 점검해보고 적용했는가?

hyeonnkang committed:
feat: #49 모임 삭제 API 생성 및 테스트

모임을 삭제 하는 API를 컨트롤러 Layer 까지 생성하고 postman으로 삭제 테스트 까지 진행했다.

현재 mapping 메소드가 DELETE로 되어있는데 추가 컬럼을 넣고 PATCH로 변경할 지는 추후 논의하고자 한다.

hyeonnkang committed:
feat: #49 모임 삭제 시 meetId 받아오는 방법 변경

기존 query string으로 모임의  id를 입력받았는데, path variable 로 변경했다.

hyeonnkang committed:
feat: #49 develop branch merge 적용 및 일부 주석 추가

develop 브랜치의 변경사항을 #49-meet-delete 브랜치로 merge 했고 MeetService 내의 getAddress 메소드의 주석을 추가해주었다.

hyeonnkang committed:
feat: #49 모임 삭제 시 인증 기능 적용

모임 삭제 시 모임을 생성한 사용자와 삭제하려는 사용자가 일치한지 확인하는 로직이 기존에 빠져있었는데 이 부분을 추가해주었다.

hyeonnkang committed:
feat: #49 Meet toString 삭제

테스트하면서 일시적으로 toString 코드를 추가했었는데 이 부분을 삭제해주었다.

Wo-ogie commented:
이거 일단 Conflict부터 해결해주세요~

hyeonnkang committed:
Merge branch 'feature/#62-meet-findById' into feature/#49-meet-delete

hyeonnkang committed:
feat: $49 MeetDeleteResponse 명세 작성

@Schema로 API 명세를 작성했다.

hyeonnkang committed:
Merge branch 'develop' into feature/#49-meet-delete

Wo-ogie commented on src/main/java/com/umc/withme/controller/MeetController.java:
```diff
@@ -82,4 +79,34 @@ public ResponseEntity<DataResponse<MeetInfoGetResponse>> getMeet(@PathVariable L
+            description = "<p><code>meetId</code>에 해당하는 모임을 삭제하고 삭제한 <code>meetId</code>를 response body에 넣어 전달합니다.</p>",
+            security = @SecurityRequirement(name = "access-token")
+    )
+    @ApiResponses({
+            @ApiResponse(responseCode = "200", description = "Success")
+    })
```
`@ApiResponses`를 설정하지 않으면 기본적으로 200 success에 대한 부분은 swagger가 자동 생성해주기 때문에 별다른 작성하신 이유가 없다면 삭제해도 될 것 같습니다.

Wo-ogie commented on src/main/java/com/umc/withme/dto/meet/MeetDeleteResponse.java:
```diff
@@ -0,0 +1,18 @@
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+
+@Getter
+@AllArgsConstructor(access = AccessLevel.PRIVATE)
+public class MeetDeleteResponse {
+
+    @Schema(example = "1", description = "삭제 성공한 모임의 id")
+    private Long meetId;
+
+    public static MeetDeleteResponse of(Long meetId) {
+        return new MeetDeleteResponse(meetId);
+    }
+}
```
삭제 성공 시 삭제한 모임의 id(PK)를 응답하는 이유가 있을까요?

hyeonnkang replied:
응답으로 필요한 데이터가 마땅히 생각나지 않아 DataResponse<>() 안에 아무것도 안넣으니 오류가 나더라구요. 
그래서 모임 id 그냥 반환하는 걸로 했는데 MeetDeleteResponse 클래스만 놔두고 meetId 는 삭제하는게 나으려나요?

hyeonnkang replied:
meetId 필드를 삭제하니까 json으로 파싱할게 없어서인지 오류가 발생해서 이 부분은 그대로 놔두고 다른 피드백들은 모두 수정사항 반영해서 푸쉬했습니다!

Wo-ogie replied:
응답할 데이터가 없는 경우에는 메타데이터만 존재하는 `BaseResponse`를 응답하시면 됩니다.

Wo-ogie commented on src/main/java/com/umc/withme/repository/MeetAddressRepository.java:
```diff
@@ -4,6 +4,7 @@
 import org.springframework.data.jpa.repository.JpaRepository;
 
 import java.util.List;
+import java.util.Optional;
```
불필요한 import문 같아 보이는데, 확인해보시고 필요 없는거면 삭제해주시면 좋을 것 같습니다.

Wo-ogie commented on src/main/java/com/umc/withme/repository/MeetMemberRepository.java:
```diff
@@ -3,5 +3,8 @@
+import java.util.List;
+
 public interface MeetMemberRepository extends JpaRepository<MeetMember, Long> {
+    List<MeetMember> findByMeet_Id(Long meetId);
```
`MeetAddressRepository`의 경우처럼 리스트 조회임을 method name에 명시해주시면 좋을 것 같아요.

Wo-ogie commented on src/main/java/com/umc/withme/service/MeetService.java:
```diff
@@ -75,4 +83,41 @@ public MeetDto findById(Long meetId) {
+     * @param meetId
+     */
+    @Transactional
+    public void deleteMeetById(Long meetId, String memberName) {
+        // 모임을 삭제하려고 하는 사용자
+        Member member = memberRepository.findByNickname(memberName)
+                .orElseThrow(NicknameNotFoundException::new);
```
`memberName`은 `email` 또는 `memberEmail`로 값과 관련 있는 변수명으로 해주시는게 좋을 것 같고, repository에서 조회할 때는 `findByEmail`을 사용해야하지 않을까요?

Wo-ogie commented on src/main/java/com/umc/withme/service/MeetService.java:
```diff
@@ -75,4 +83,41 @@ public MeetDto findById(Long meetId) {
+        Member member = memberRepository.findByNickname(memberName)
+                .orElseThrow(NicknameNotFoundException::new);
+
+        // 삭제하려는 모임
+        Meet meet = meetRepository.findById(meetId)
+                .orElseGet(null);
```
삭제하려는 모임이 없을 때 `meet`에 `null`을 넣는 이유가 있나요? 특별한 이유가 없다면 별로 좋지 않아보입니다.

Wo-ogie commented on src/main/java/com/umc/withme/service/MeetService.java:
```diff
@@ -75,4 +83,41 @@ public MeetDto findById(Long meetId) {
+        Meet meet = meetRepository.findById(meetId)
+                .orElseGet(null);
+
+        // 모임의 주인
+        Member meetLeader = memberRepository.findById(meet.getCreatedBy())
+                .orElseThrow(UnauthorizedException::new);
```
여기는 왜 `UnauthorizedException`을 던지는걸까요? 해당 예외와 관련 없어 보이는 상황같은데...

Wo-ogie commented on src/main/java/com/umc/withme/service/MeetService.java:
```diff
@@ -75,4 +83,41 @@ public MeetDto findById(Long meetId) {
+                .orElseThrow(UnauthorizedException::new);
+
+        // 모임의 주인과 사용자가 일치하면 해당 모임 삭제. 일치하지 않으면 예외 발생
+        if(meetLeader.equals(member)){
+            meetAddressRepository.findAllByMeet_Id(meetId)
+                    .stream()
+                    .forEach(ma -> meetAddressRepository.delete(ma));
+
+            meetMemberRepository.findByMeet_Id(meetId)
+                    .stream()
+                    .forEach(mm -> meetMemberRepository.delete(mm));
+
+            meetRepository.delete(
+                    meetRepository.findById(meetId)
+                            .orElseThrow(EntityNotFoundException::new)
+            );
+        }else
+            throw new DeleteBadRequestException(meet.getId(), memberName);
+    }
```
1. 다음과 같이 하시면 `if-else` 구문에서 `if`문으로 간략화 할 수 있으며, 조건문의 볼륨도 작아집니다.
```java
if (!meetLeader.equals(member)) {
    // throw new 관련 예외;
}
// 정상 로직 수행
```

2. 위에서 한 번 조회했던 모임이 있는데 모임을 삭제할 때 굳이 또 조회해야 할까요?

3. 예외의 naming을 `DeleteBadRequestException`으로 하는 것이 좋을지 조금 고민이 되는 부분입니다. 정말 client에서 **요청을 잘못 한 것**이 맞을까요? 이 경우에는 요청 방법이 잘못되거나 한 경우라기 보다는 **특정 resource를 수정/삭제할 권한이 없는 상황**으로 보는 것이 더 적합할 것 같습니다. 또한 custom exception의 naming에 모임 관련된 예외임을 드러내주는 것이 좋을 것 같고, `ExceptionType`에 error message도 이러한 내용들이 잘 작성되었으면 좋겠습니다.

4. Custom exception을 정의하셨으면 `ExceptionType`에 관련 내용도 **반드시 작성해주세요**. 이 작업을 하지 않으면 추후 client에서는 어떤 상황이든지 "알 수 없는 에러가 발생했습니다"만 보게 되어 뭐 때문에 문제가 되는 상황인지 파악하기가 어려워집니다.

5. `.stream()` 안쓰셔도 됩니다.

hyeonnkang committed:
feat: #49 코드리뷰 반영 (모임삭제 response 제외)

- MeetController 내의 @ApiResponse 수정
- MeetAddressRepository 내의 불필요한 import 문 삭제
- MeetMemberRepository 내의 findByMeet_id 결과값이 리스트임을 강조하도록 All 키워드 추가
- MeetService의 deleteMeetById 메소드 내의 memberName을 memberEmail 로 네이밍 변경
- 모임 조회 시 null 로 넣어주던 것을 exception throw 로 교체
- 모임 주인 조회 시 unauthorized exception -> MemberIdNotFoundException으로 교체
- DeleteBadRequestException -> MeetDeleteUnautorizedException 으로 네이밍 변경

모임 삭제 response는 클래스 내 id를 삭제하니 클래스 내의 필드가 하나도 없어서 오류가 발생하는 것 같습니다. 반환값을 무엇으로 할 지 상의가 필요할 것 같습니다!

Wo-ogie commented on src/main/java/com/umc/withme/service/MeetService.java:
```diff
@@ -87,37 +88,33 @@ public MeetDto findById(Long meetId) {
         // 모임을 삭제하려고 하는 사용자
-        Member member = memberRepository.findByNickname(memberName)
+        Member member = memberRepository.findByEmail(memberEmail)
                 .orElseThrow(NicknameNotFoundException::new);
```
이메일로 조회하는 상황이니까 예외도 적절하게 바꿔주세요

Wo-ogie commented on src/main/java/com/umc/withme/controller/MeetController.java:
```diff
@@ -67,6 +65,7 @@ public ResponseEntity<DataResponse<MeetCreateResponse>> createMeet(
     )
     @ApiResponses({
             @ApiResponse(responseCode = "200", description = "Success"),
+            @ApiResponse(responseCode = "404", description = "2400: <code>id</code>에 해당하는 모임이 없는 경우", content = @Content)
     })
     @GetMapping("/meets/{meetId}")
```
API 명세는 프론트엔드 개발자분들이 보게 될 문서니까, 이 내용을 고려해서 조금 더 신경써주시면 좋을 것 같습니다.
현재 프론트엔드에서 API를 호출할 때 모임의 PK 값을 전달하는데 이 이름이 지금 `meetId`이므로 API 명세에도 `meetId`라고 명시해주는 게 더 명확할 것 같아요. `id`라고만 적어도 대충 알아는 듣겠지만, 좀 더 친절하게 작성해주면 더 좋겠죠?

Wo-ogie commented on src/main/java/com/umc/withme/exception/ExceptionType.java:
```diff
@@ -43,8 +44,10 @@ public enum ExceptionType {
+    MEMBER_ID_NOT_FOUND_EXCEPTION(1402, "해당 id를 갖는 회원을 찾을 수 없습니다.", MemberIdNotFoundException.class),
 
     // Meet
+    MEET_DELETE_UNAUTHORIZED_EXCEPTION(2100, "해당 모임을 삭제할 권한이 없습니다.", MeetIdNotFoundException.class),
```
여기 수정해주세요.

Wo-ogie commented on src/main/java/com/umc/withme/exception/meet/MeetDeleteUnauthorizedException.java:
```diff
@@ -0,0 +1,17 @@
+
+import com.umc.withme.exception.common.UnauthorizedException;
+
+public class MeetDeleteUnauthorizedException extends UnauthorizedException {
```
이 예외의 경우 unauthorized 보다는 forbidden이 적합하지 않을까요?

hyeonnkang replied:
넵 말씀하신 부분 전부 고쳐서 푸쉬했습니다! 감사합니다!

Wo-ogie commented:
Controller class type level에 `@Tag` annotation 활용해서 제목도 간단하게 붙여주시면 좋을 것 같아요.

hyeonnkang committed:
feat: #49 코드리뷰 반영

- 이메일로 멤버 조회 시 예외 throw 적절하게 변경
- API 명세 id -> meetId로 변경
- ExceptionType에서 exception class 적절하게 변경
- 모임 삭제 시 삭제 권한이 없을 경우 MeetDeleteForbiddenException을 throw 하는 것으로 변경

Wo-ogie commented on src/main/java/com/umc/withme/exception/ExceptionType.java:
```diff
@@ -47,7 +48,7 @@ public enum ExceptionType {
 
     // Meet
-    MEET_DELETE_UNAUTHORIZED_EXCEPTION(2100, "해당 모임을 삭제할 권한이 없습니다.", MeetIdNotFoundException.class),
+    MEET_DELETE_FORBIDDEN_EXCEPTION(2300, "해당 모임을 삭제할 권한이 없습니다.", MeetDeleteForbiddenException.class),
```
Forbidden은 403에러니까 저희 회의 때 얘기한대로면 에러 코드가 2401이 될 것 같습니다.

hyeonnkang replied:
넵 변경했습니다!

hyeonnkang committed:
feat: #49 코드리뷰 - 에러코드 변경

MEET_DELETE_FORBIDDEN_EXCEPTION 에러 코드가 기존에 2300 이었는데 Forbidden 에러 코드가 403 이므로 2401로 변경했다.

hyeonnkang committed:
feat: #49 주석 추가

param 주석 빠진 곳이 있어서 추가해주었습니다.

Wo-ogie commented on src/main/java/com/umc/withme/service/MeetService.java:
```diff
@@ -75,4 +83,37 @@ public MeetDto findById(Long meetId) {
+     * @param meetId 삭제할 모임의 id
+     */
+    @Transactional
+    public void deleteMeetById(Long meetId, String memberEmail) {
+        // 모임을 삭제하려고 하는 사용자
+        Member member = memberRepository.findByEmail(memberEmail)
+                .orElseThrow(EmailNotFoundException::new);
+
+        // 삭제하려는 모임
+        Meet meet = meetRepository.findById(meetId)
+                .orElseThrow(() -> new MeetIdNotFoundException(meetId));
+
+        // 모임의 주인
+        Member meetLeader = memberRepository.findById(meet.getCreatedBy())
+                .orElseThrow(() -> new MemberIdNotFoundException(meet.getCreatedBy()));
+
+        // 모임의 주인과 사용자가 일치하면 해당 모임 삭제. 일치하지 않으면 예외 발생
+        if (!meetLeader.equals(member))
+            throw new MeetDeleteForbiddenException(meet.getId(), memberEmail);
```
이거 비교를 다음과 같이 PK로 하면 `meetLeader`를 조회하는 코드가 없어도 될 것 같아요. 그럼 코드가 더 간결해질 것 같습니다. 어떨까요?
```java
Long meetLeaderId = meet.getCreatedBy();

if (meetLeaderId.equals(member.getId()) {
    throw 예외;
}
```

심지어는 `deleteMeetById()`가 전달받는 argument가 `String memberEmail`이 아닌 `Long loginMemberId`라면 "모임을 삭제하려고 하는 사용자"조차 조회할 필요 없이 다음과 같이 검증이 가능할 것 같습니다.
```java
Long meetLeaderId = meet.getCreatedBy();

if (!loginMemberId.equals(meetLeaderId)) {
    throw 예외
}
```
참고로 로그인한 사용자의 PK는 controller method에서 `principle.getMemberId()`로 받을 수 있습니다.

hyeonnkang replied:
넵 말씀해주신 부분 확인하고 수정했습니다! 감사합니다!

hyeonnkang committed:
feat: #49 코드리뷰

코드 리뷰를 바탕으로 delete 사용자 검증 시 사용자 pk 를 비교하는 방법으로 변경해 조회 쿼리를 줄였다.

Wo-ogie reviewed (approved):
조금 천천히 해도 괜찮으니, 다음에 작업하실 때도 꼼꼼히 코드리뷰 내용들 고려하시면서 하시면 더 좋을 것 같습니다. 😊
수고하셨습니다!
